.repo/sample-list-generator/src/libs/api-detector.ts

```diff
@@ -0,0 +1,201 @@
+  sampleFolderPath: string
+): Promise<ApiItemWithType[]> => {
+  // get all js files in the folder
+  const jsFiles = (await getAllFiles(sampleFolderPath)).filter((file) =>
+    file.endsWith('.js')
+  );
+
+  const calls: ApiItemWithType[] = [];
+
+  await Promise.all(
+    jsFiles.map(async (file) => {
+      const callsFromFile = await extractApiCalls(await fs.readFile(file));
+      calls.push(...callsFromFile);
+    })
+  );
+
+  return uniqueItems(calls);
+};
+
+/**
+ * Gets the complete API call for the member expression.
+ * @param path - The path to the MemberExpression node.
+ * @returns The full member expression.
+ * @example
+ * getFullMemberExpression(path.node)
+ * // returns ['chrome', 'tabs', 'query']
+ */
+export function getFullMemberExpression(
+  path: babel.NodePath<babel.types.MemberExpression>
+): string[] {
+  const result: string[] = [];
+
+  // Include the chrome. or browser. identifier
+  if (isIdentifier(path.node.object)) {
+    result.push(path.node.object.name);
+  } else {
+    // We don't support expressions
+    return result;
+  }
+
+  while (path) {
+    if (isIdentifier(path.node.property)) {
+      result.push(path.node.property.name);
+    } else {
+      // We don't support expressions
+      break;
+    }
+
+    const parentPath = path.parentPath;
+
+    if (!parentPath || !parentPath.isMemberExpression()) {
+      break;
+    } else {
+      path = parentPath;
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Gets the namespace and property name of an api call.
+ * @param parts - The parts of the api call.
+ * @returns The namespace and property name of the api call.
+ * @example
+ * getApiItem(['chrome', 'tabs', 'query'])
+ * // returns { namespace: 'tabs', propertyName: 'query' }
+ * getApiItem(['chrome', 'devtools', 'inspectedWindow', 'eval'])
+ * // returns { namespace: 'devtools.inspectedWindow', propertyName: 'eval' }
+ */
+export function getApiItem(parts: string[]): ApiItem {
+  let namespace = '';
+  let propertyName = '';
+
+  // For some apis like `chrome.devtools.inspectedWindow.eval`,
+  // the namespace is actually `devtools.inspectedWindow`.
+  // So we need to check if the first two parts combined is a valid namespace.
+  if (EXTENSION_API_MAP[`${parts[0]}.${parts[1]}`]) {
+    namespace = `${parts[0]}.${parts[1]}`;
+    propertyName = parts[2];
+  } else {
+    namespace = parts[0];
+    propertyName = parts[1];
+  }
+
+  return { namespace, propertyName };
+}
+
+/**
+ * Filters an array of ApiItemWithType to remove duplicates.
+ * @param array - The array of ApiItemWithType to filter.
+ */
+function uniqueItems(array: ApiItemWithType[]) {
+  const tmp = new Set<string>();
+  return array.filter((item) => {
+    const fullApiString = `${item.namespace}.${item.propertyName}`;
+    return !tmp.has(fullApiString) && tmp.add(fullApiString);
+  });
+}
+
+/**
+ * Extracts all chrome and browser api calls from a file.
+ * @param file - The file to extract api calls from.
```

Comment:
nit: why is this not named buffer? 

I think it'd be better to change the method to accept a string instead of a buffer this moves responsibility to the caller.

Reply:
Done. https://github.com/GoogleChrome/chrome-extensions-samples/pull/963/commits/4fcffeefe325e859cf97af19f0b8af87c3cdff5f & https://github.com/GoogleChrome/chrome-extensions-samples/pull/963/commits/0c663a5dbfb4efc4b48de28dd822a6a9f1aea504